### ISSUE TITLE
DDPG算法Actor训练一次后损失一直是0

### ISSUE THREAD

**Issue by zp7541** [PaddlePaddle/Paddle#29110]
`class ActorModel(parl.Model):
    def __init__(self, act_dim):
        self.conv1 = layers.conv2d(num_filters=16, filter_size=5, stride=2, act='relu')
        self.conv2 = layers.conv2d(num_filters=32, filter_size=3, stride=2, act='relu')
        self.conv3 = layers.conv2d(num_filters=32, filter_size=3, stride=1, act='relu')
        self.conv4 = layers.conv2d(num_filters=64, filter_size=3, stride=1, act='relu')
        self.conv5 = layers.conv2d(num_filters=128, filter_size=3, stride=1)
        self.fc1 = layers.fc(200, act='relu')
        self.fc2 = layers.fc(200, act='relu')
        self.fc3 = layers.fc(3, act='tanh')

    def policy(self, obs):
        conv1 = self.conv1(obs)
        # print('conv1.shape', conv1.shape)
        conv2 = self.conv2(conv1)
        conv3 = self.conv3(conv2)
        conv4 = self.conv4(conv3)
        conv5 = self.conv5(conv4)

        flatten = layers.flatten(conv5, axis=1)

        fc1 = self.fc1(flatten)
        fc2 = self.fc2(fc1)
        fc3 = self.fc3(fc2)

        return fc3

class CriticModel(parl.Model):
    def __init__(self):
        self.conv1 = layers.conv2d(num_filters=16, filter_size=5, stride=2, act='relu')
        self.conv2 = layers.conv2d(num_filters=32, filter_size=3, stride=2, act='relu')
        self.conv3 = layers.conv2d(num_filters=32, filter_size=3, stride=1, act='relu')
        self.conv4 = layers.conv2d(num_filters=64, filter_size=3, stride=1, act='relu')
        # self.conv5 = layers.conv2d(num_filters=128, filter_size=3, stride=1)
        self.fc1 = layers.fc(200, act='relu')
        self.fc2 = layers.fc(200, act='relu')
        self.fc3 = layers.fc(1, act='relu')

    def value(self, obs, act):
        conv1 = self.conv1(obs)
        conv2 = self.conv2(conv1)
        conv3 = self.conv3(conv2)
        conv4 = self.conv4(conv3)
        # conv5 = self.conv5(conv4)

        concat = layers.concat([layers.flatten(conv4, axis=1), act], axis=1)
        fc1 = self.fc1(concat)
        fc2 = self.fc2(fc1)
        fc3 = self.fc3(fc2)

        return fc3

class Model(parl.Model):
    def __init__(self, act_dim):
        self.actor_model = ActorModel(act_dim)
        self.critic_model = CriticModel()

    def policy(self, obs):
        return self.actor_model.policy(obs)

    def value(self, obs, act):
        return self.critic_model.value(obs, act)

    def get_actor_params(self):
        return self.actor_model.parameters()

class DDPG(parl.Algorithm):
    def __init__(self,
                 model,
                 gamma=None,
                 tau=None,
                 actor_lr=None,
                 critic_lr=None):
        """  DDPG algorithm

        Args:
            model (parl.Model): forward network of actor and critic.
                                The function get_actor_params() of model should be implemented.
            gamma (float): discounted factor for reward computation.
            tau (float): decay coefficient when updating the weights of self.target_model with self.model
            actor_lr (float): learning rate of the actor model
            critic_lr (float): learning rate of the critic model
        """
        assert isinstance(gamma, float)
        assert isinstance(tau, float)
        assert isinstance(actor_lr, float)
        assert isinstance(critic_lr, float)
        self.gamma = gamma
        self.tau = tau
        self.actor_lr = actor_lr
        self.critic_lr = critic_lr

        self.model = model
        self.target_model = deepcopy(model)

    def predict(self, obs):
        """ use actor model of self.model to predict the action
        """
        return self.model.policy(obs)

    def learn(self, obs, action, reward, next_obs, terminal):
        """ update actor and critic model with DDPG algorithm
        """
        actor_cost = self._actor_learn(obs)
        critic_cost = self._critic_learn(obs, action, reward, next_obs,
                                         terminal)
        return actor_cost, critic_cost

    def _actor_learn(self, obs):
        action = self.model.policy(obs)
        Q = self.model.value(obs, action)
        cost = layers.reduce_mean(-1.0 * Q)
        optimizer = fluid.optimizer.AdamOptimizer(self.actor_lr)
        optimizer.minimize(cost, parameter_list=self.model.get_actor_params())
        return cost

    def _critic_learn(self, obs, action, reward, next_obs, terminal):
        next_action = self.target_model.policy(next_obs)
        next_Q = self.target_model.value(next_obs, next_action)

        terminal = layers.cast(terminal, dtype='float32')
        target_Q = reward + (1.0 - terminal) * self.gamma * next_Q
        target_Q.stop_gradient = True

        Q = self.model.value(obs, action)
        cost = layers.square_error_cost(Q, target_Q)
        cost = layers.reduce_mean(cost)
        optimizer = fluid.optimizer.AdamOptimizer(self.critic_lr)
        optimizer.minimize(cost)
        return cost

    def sync_target(self, decay=None, share_vars_parallel_executor=None):
        if decay is None:
            decay = 1.0 - self.tau
        self.model.sync_weights_to(
            self.target_model,
            decay=decay,
            share_vars_parallel_executor=share_vars_parallel_executor)

class Agent(parl.Agent):
    def __init__(self, algorithm, obs_dim, act_dim=4):

        self.obs_dim = obs_dim
        self.act_dim = act_dim
        super(Agent, self).__init__(algorithm)

        # 注意，在最开始的时候，先完全同步target_model和model的参数
        self.alg.sync_target(decay=0)

    def build_program(self):
        self.pred_program = fluid.Program()
        self.learn_program = fluid.Program()

        with fluid.program_guard(self.pred_program):
            obs = layers.data(
                name='obs', shape=self.obs_dim, dtype='float32')
            self.pred_act = self.alg.predict(obs)

        with fluid.program_guard(self.learn_program):
            obs = layers.data(
                name='obs', shape=self.obs_dim, dtype='float32')
            act = layers.data(
                name='act', shape=[self.act_dim], dtype='float32')
            reward = layers.data(name='reward', shape=[], dtype='float32')
            next_obs = layers.data(
                name='next_obs', shape=self.obs_dim, dtype='float32')
            terminal = layers.data(name='terminal', shape=[], dtype='bool')
            self.actor_cost, self.critic_cost = self.alg.learn(obs, act, reward, next_obs,
                                                 terminal)

    def predict(self, obs):
        obs = np.expand_dims(obs, axis=0)
        act = self.fluid_executor.run(
            self.pred_program, feed={'obs': obs},
            fetch_list=[self.pred_act])[0]
        return act

    def learn(self, obs, act, reward, next_obs, terminal):
        feed = {
            'obs': obs,
            'act': act,
            'reward': reward,
            'next_obs': next_obs,
            'terminal': terminal
        }
        actor_cost, critic_cost = self.fluid_executor.run(
            self.learn_program, feed=feed, fetch_list=[self.actor_cost, self.critic_cost])
        self.alg.sync_target()
        return actor_cost, critic_cost

class ReplayMemory(object):
    def __init__(self, max_size):
        self.buffer = collections.deque(maxlen=max_size)

    def append(self, exp):
        self.buffer.append(exp)

    def sample(self, batch_size):
        mini_batch = random.sample(self.buffer, batch_size)
        obs_batch, action_batch, reward_batch, next_obs_batch, done_batch = [], [], [], [], []

        for experience in mini_batch:
            s, a, r, s_p, done = experience
            obs_batch.append(s)
            action_batch.append(a)
            reward_batch.append(r)
            next_obs_batch.append(s_p)
            done_batch.append(done)

        return np.array(obs_batch).astype('float32'), \
            np.array(action_batch).astype('float32'), np.array(reward_batch).astype('float32'),\
            np.array(next_obs_batch).astype('float32'), np.array(done_batch).astype('float32')

    def __len__(self):
        return len(self.buffer)

# ----------------------------------------- ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ -----------------------------------------------------------------

# ----------------------------↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓ hyper parameters ↓ ↓ ↓ ↓ ↓ ↓ ↓ ↓--------------------------------------------------------------
ACTOR_LR = 1e-6  # Actor网络的 learning rate
CRITIC_LR = 1e-5  # Critic网络的 learning rate

GAMMA = 0.99      # reward 的衰减因子
TAU = 0.001       # 软更新的系数
MEMORY_SIZE = int(1e6)                  # 经验池大小
MEMORY_WARMUP_SIZE = MEMORY_SIZE // 20  # 预存一部分经验之后再开始训练
BATCH_SIZE = 16
REWARD_SCALE = 0.1   # reward 缩放系数
NOISE = 0.05         # 动作噪声方差

MAX_EPISODES = 60000  # 训练的总episode数
# ----------------------------------------- ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ ↑ -----------------------------------------------------------------

ACT_DIM = 3
OBS_DIM = [None, 3, 200, 360]
USE_CARLA = True
NORMALIZE_OBS = False
MEAN = [0.485, 0.456, 0.406]
STD = [0.229, 0.224, 0.225]

model = Model(ACT_DIM)
algorithm = DDPG(model, gamma=GAMMA, tau=TAU, actor_lr=ACTOR_LR, critic_lr=CRITIC_LR)
agent = Agent(algorithm, OBS_DIM, ACT_DIM)
rpm = ReplayMemory(MEMORY_SIZE)`

**Comment by yinhaofeng** [comment]
你好，能给一下可以复现的完整代码，数据和环境吗？

**Comment by zp7541** [comment]
> 你好，能给一下可以复现的完整代码，数据和环境吗？

已经弄好了,请问下parl.layers.batch_norm这个东西怎么用啊，有参考示例吗

**Comment by yinhaofeng** [comment]
https://www.paddlepaddle.org.cn/documentation/docs/zh/api_cn/layers_cn/batch_norm_cn.html#batch-norm

**Comment by zp7541** [comment]
`class ActorModel(parl.Model):
    def __init__(self, act_dim):

        self.conv1 = layers.conv2d(num_filters=16, filter_size=5, stride=2)
        self.bn1 = layers.batch_norm(act='relu')

        self.conv2 = layers.conv2d(num_filters=32, filter_size=3, stride=2)
        self.bn2 = layers.batch_norm(act='relu')

        self.conv3 = layers.conv2d(num_filters=32, filter_size=3, stride=1)
        self.bn3 = layers.batch_norm(act='relu')

        self.conv4 = layers.conv2d(num_filters=64, filter_size=3, stride=1)
        self.bn4 = layers.batch_norm(act='relu')
        # self.conv5 = layers.conv2d(num_filters=128, filter_size=3, stride=1)

        self.fc1 = layers.fc(200, act='relu')
        self.fc2 = layers.fc(200, act='relu')
        self.fc3 = layers.fc(2, act='tanh')

    def policy(self, obs):
        conv1 = self.conv1(obs)
        bn1 = self.bn1(conv1)
        # bn1 = layers.batch_norm(conv1)
        # print('conv1.shape', conv1.shape)

        conv2 = self.conv2(bn1)
        bn2 = self.bn2(conv2)

        conv3 = self.conv3(bn2)
        bn3 = self.bn3(conv3)

        conv4 = self.conv4(bn3)
        bn4 = self.bn4(conv4)
        # conv5 = self.conv5(conv4)

        flatten = layers.flatten(bn4, axis=1)

        fc1 = self.fc1(flatten)
        fc2 = self.fc2(fc1)
        fc3 = self.fc3(fc2)

        return fc3
`

这个报错TypeError: batch_norm() got an unexpected keyword argument 'fuse_with_relu'， 请问下应该怎么改啊

**Comment by yinhaofeng** [comment]
你看一下参数啊，首先你要有input，表明bn层在哪一层后面

**Comment by zp7541** [comment]
> 你看一下参数啊，首先你要有input，表明bn层在哪一层后面

我后面有加啊， 动图的不也大概是这个方式写的吗。 能给个dqn加batch_norm的代码吗

**Comment by yinhaofeng** [comment]
https://github.com/PaddlePaddle/models/search?q=batch_norm  可以看看这个
